src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,955 @@
+import net.minecraft.util.thread.ReentrantBlockableEventLoop;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.config.ConfigTracker;
+import net.neoforged.neoforge.internal.versions.neoforge.NeoForgeVersion;
+import net.neoforged.neoforge.network.connection.ConnectionUtils;
+import net.neoforged.neoforge.network.event.RegisterPayloadHandlerEvent;
+import net.neoforged.neoforge.network.filters.NetworkFilters;
+import net.neoforged.neoforge.network.handling.ConfigurationPayloadContext;
+import net.neoforged.neoforge.network.handling.IPacketHandler;
+import net.neoforged.neoforge.network.handling.ISynchronizedWorkHandler;
+import net.neoforged.neoforge.network.handling.PlayPayloadContext;
+import net.neoforged.neoforge.network.negotiation.NegotiableNetworkComponent;
+import net.neoforged.neoforge.network.negotiation.NegotiationResult;
+import net.neoforged.neoforge.network.negotiation.NetworkComponentNegotiator;
+import net.neoforged.neoforge.network.payload.ModdedNetworkComponent;
+import net.neoforged.neoforge.network.payload.ModdedNetworkPayload;
+import net.neoforged.neoforge.network.payload.ModdedNetworkQueryComponent;
+import net.neoforged.neoforge.network.payload.ModdedNetworkQueryPayload;
+import net.neoforged.neoforge.network.payload.ModdedNetworkSetupFailedPayload;
+import org.apache.commons.compress.utils.Lists;
+import org.jetbrains.annotations.ApiStatus;
+import org.slf4j.Logger;
+
+/**
+ * Defines the registry for all modded network packets.
+ * <p>
+ * This registry is responsible for storing all known modded network packets, and for handling the negotiation of modded network packets between the client and the server.
+ * </p>
+ * <p>
+ * Additionally, this registry is responsible for handling all modded network packets that are not natively known once they arrive at the receiving end.
+ * </p>
+ * <p>
+ * To prevent payloads from being send to a client that has no idea what to do with them, the registry provides endpoints for the vanilla code base to check if a packet can be send to a client.
+ * </p>
+ */
+@ApiStatus.Internal
+public class NetworkRegistry {
+
+    private static final Logger LOGGER = LogUtils.getLogger();
+
+    private static final AttributeKey<NetworkPayloadSetup> ATTRIBUTE_PAYLOAD_SETUP = AttributeKey.valueOf("neoforge:payload_setup");
+    private static final AttributeKey<Boolean> ATTRIBUTE_IS_MODDED_CONNECTION = AttributeKey.valueOf("neoforge:is_modded_connection");
+    private static final AttributeKey<PacketFlow> ATTRIBUTE_FLOW = AttributeKey.valueOf("neoforge:flow");
+
+    private static final NetworkRegistry INSTANCE = new NetworkRegistry();
+
+    public static NetworkRegistry getInstance() {
+        return INSTANCE;
+    }
+
+    private boolean setup = false;
+    private final Map<ResourceLocation, ConfigurationRegistration<?>> knownConfigurationRegistrations = Maps.newHashMap();
+    private final Map<ResourceLocation, PlayRegistration<?>> knownPlayRegistrations = Maps.newHashMap();
```

Comment:
Why do these use `Maps.newHashMap()` and other parts of this class use `new HashMap<>()`?

Reply:
Personal preference.
I tend to do that over calling a constructor because it is better unit-testable in the end.

Reply:
Note that the guava method is effectively deprecated, and should not be used in new code:
![image](https://github.com/neoforged/NeoForge/assets/13494793/095d091f-66ce-4400-ae97-ff6470493d05)